pkg/handler/category.go

```diff
@@ -52,19 +53,55 @@ func CreateCategory(c *fiber.Ctx) {
 func UpdateCategory(c *fiber.Ctx) {
 	id := c.Params("id")
 	db := database.DB
-	form := new(model.Category)
-	if err := c.BodyParser(&form); err != nil {
-		c.Status(500).JSON(fiber.Map{"status": "error", "message": "Couldn't update category", "data": nil})
-		return
-	}
+	// Get the category with given id
+	// Set updated fields
+	// Persist the end result
+
 	var category model.Category
-	db.First(&category, id)
-	if category.ID == 0 {
+	db.Preload("Platforms").First(&category, id)
+	switch {
+	case db.Error != nil:
+		c.Status(503)
+		return
+	case category.ID == 0:
 		c.Status(404).JSON(fiber.Map{"status": "error", "message": "No category found with ID", "data": nil})
 		return
 	}
+
+	form := new(model.CategoryCreateUpdate)
+	if err := c.BodyParser(&form); err != nil {
+		c.Status(500).JSON(fiber.Map{"status": "error", "message": "Couldn't update category", "data": nil})
+		return
+	}
+
 	category.Name = form.Name
 	category.Description = form.Description
+
+	// filter the platforms coming from the update body
+	// remove association for those which are filtered out, removed with the update operation
+	platforms := make([]model.Platform, len(form.PlatformNames))
+	for i, p := range form.PlatformNames {
+		platforms[i] = model.Platform{Name: p}
+	}
+
+	oldPlatforms := category.Platforms
+
+	filteredPlatforms := filterToCreate(oldPlatforms, platforms)
+	deletePlatforms := filterToDelete(oldPlatforms, platforms)
+
+	// todo: find a way to not replace existing associations when you're not changing the platforms at all
+	// keep in mind that you are creating new platform instances each time update function's triggered -> Platform{Name: p}
+	//db.Model(&category).Association("Platforms").Replace(platforms)
```

Comment:
This Replace method causes platform records to get overwritten each time. So I added 2 functions to filter platforms when updating categories.
User wants to update a category and picks which platforms to add or remove from the category.

Reply:
Your solution looks feasible. Thank you for effort